drivers/pinctrl/pinctrl_esp32.c

```diff
@@ -0,0 +1,270 @@
+{
+	return ((BIT(pin) & SOC_GPIO_VALID_OUTPUT_GPIO_MASK) != 0);
+}
+
+static int esp32_pin_apply_config(const struct device *dev, uint32_t pin, uint32_t flags)
+{
+	gpio_dev_t *const gpio_base = (gpio_dev_t *)DT_REG_ADDR(DT_NODELABEL(gpio0));
+	uint32_t io_pin = (uint32_t) pin + ((ESP32_PORT_IDX(pin) == 1 && pin < 32) ? 32 : 0);
+	int ret = 0;
+
+	if (!esp32_pin_is_valid(io_pin)) {
+		return -EINVAL;
+	}
+
+#if SOC_RTCIO_INPUT_OUTPUT_SUPPORTED
+	if (rtc_gpio_is_valid_gpio(io_pin)) {
+		rtcio_hal_function_select(rtc_io_num_map[io_pin], RTCIO_FUNC_DIGITAL);
+	}
+#endif
+
+	if (io_pin >= GPIO_NUM_MAX) {
+		ret = -EINVAL;
+		goto end;
+	}
+
+	/* Set pin function as GPIO */
+	gpio_ll_iomux_func_sel(GPIO_PIN_MUX_REG[io_pin], PIN_FUNC_GPIO);
+
+	if (flags & ESP32_PULL_UP_FLAG) {
+		if (!rtc_gpio_is_valid_gpio(io_pin) || SOC_GPIO_SUPPORT_RTC_INDEPENDENT) {
+			gpio_ll_pulldown_dis(&GPIO, io_pin);
+			gpio_ll_pullup_en(&GPIO, io_pin);
+		} else {
+#if SOC_RTCIO_INPUT_OUTPUT_SUPPORTED
+			int rtcio_num = rtc_io_num_map[io_pin];
+
+			rtcio_hal_pulldown_disable(rtc_io_num_map[io_pin]);
+
+			if (rtc_io_desc[rtcio_num].pullup) {
+				rtcio_hal_pullup_enable(rtc_io_num_map[io_pin]);
+			} else {
+				ret = -ENOTSUP;
+				goto end;
+			}
+#endif
+		}
+	} else if (flags & ESP32_PULL_DOWN_FLAG) {
+		if (!rtc_gpio_is_valid_gpio(io_pin) || SOC_GPIO_SUPPORT_RTC_INDEPENDENT) {
+			gpio_ll_pullup_dis(&GPIO, io_pin);
+			gpio_ll_pulldown_en(&GPIO, io_pin);
+		} else {
+#if SOC_RTCIO_INPUT_OUTPUT_SUPPORTED
+			int rtcio_num = rtc_io_num_map[io_pin];
+
+			rtcio_hal_pulldown_enable(rtc_io_num_map[io_pin]);
+
+			if (rtc_io_desc[rtcio_num].pullup) {
+				rtcio_hal_pullup_disable(rtc_io_num_map[io_pin]);
+			} else {
+				ret = -ENOTSUP;
+				goto end;
+			}
+#endif
+		}
+	}
+
+	if (flags & ESP32_DIR_OUT_FLAG) {
+		if (!esp32_pin_is_output_capable(pin)) {
+			ret = -EINVAL;
+			goto end;
+		}
+
+		if (flags & ESP32_OPEN_DRAIN_FLAG) {
+			gpio_ll_od_enable(gpio_base, io_pin);
+		} else {
+			gpio_ll_od_disable(gpio_base, io_pin);
+		}
+
+		/* Set output pin initial value */
+		if (flags & ESP32_DIR_OUT_HIGH_FLAG) {
+			gpio_ll_set_level(gpio_base, io_pin, 1);
+		} else if (flags & ESP32_DIR_OUT_LOW_FLAG) {
+			gpio_ll_set_level(gpio_base, io_pin, 0);
+		}
+
+		gpio_ll_output_enable(&GPIO, io_pin);
+		esp_rom_gpio_matrix_out(io_pin, SIG_GPIO_OUT_IDX, false, false);
+	}
+
+	if (flags & ESP32_DIR_INP_FLAG) {
+		gpio_ll_input_enable(&GPIO, io_pin);
+	}
+
+end:
+	return ret;
+}
+
+static int esp32_pin_configure(const uint32_t pin_mux, const uint32_t pin_cfg)
+{
+	const struct device *port_device;
+	uint32_t pin_num = ESP32_PIN_NUM(pin_mux);
+	uint32_t sig_in = ESP32_PIN_SIGI(pin_mux);
+	uint32_t sig_out = ESP32_PIN_SIGO(pin_mux);
+	uint32_t flags = 0;
+
+	if (ESP32_PORT_IDX(pin_num) >= gpio_ports_cnt) {
+		return -EINVAL;
+	}
+
+	port_device = gpio_ports[ESP32_PORT_IDX(pin_num)];
+
+	if ((port_device == NULL) || (!device_is_ready(port_device))) {
+		return -ENODEV;
+	}
```

Comment:
this driver continues depending on GPIO, please make driver independent of GPIO. See e.g. GD32 driver where GPIO information is used from DT, but no GPIO device is used internally.

Reply:
Fixed.